### PR TITLE
Added UPF Support

### PR DESCRIPTION
**Pull request by derrickqi2003**
Added UPF support to hammer. Generates a single power domain UPF. It should do exactly what hasCPFsupport does. Tested it by running PAR on the tapeout vlsi lab and manually feeding into innovus with the innovus command below:
read_power_intent -1801 power_spec.upf
[power_spec.pdf](https://github.com/ucb-bar/hammer/files/11116390/power_spec.pdf)
.

**Related PRs / Issues**

**Type of change**:
- [ ] Bug fix
- [X] New feature
- [ ] Other enhancement

**Impact**:
- [X] Change to core Hammer
- [ ] Change to a Hammer plugin
- [ ] Other

**Contributor Checklist**:
- [X] Did you set `master` as the base branch?
- [X] Did you state the type-of-change/impact?
- [X] Did you delete any extraneous prints/debugging code?
- [ ] (If applicable) Did you add documentation for the feature?
- [ ] (If applicable) Did you update the `poetry.lock` file if you updated the requirements in `pyproject.toml`?
- [ ] (If applicable) Did you add a unit test demonstrating the PR?
- [ ] (If applicable) Did you run this through the e2e integration tests?
- [ ] (If applicable) Did you update the submodules in `e2e/` if this feature depends on updated plugins?

### CODE REVIEW

File: hammer/vlsi/hammer_vlsi_impl.py [hunk 1]
```diff
@@ -2059,11 +2059,52 @@ def sdf_file(self, value: Optional[str]) -> None:
...
+        output.append("\t-primary_ground_net {g}".format(g=ground_nets[0].name))
+        #Add Port States 
+        for p_net in power_nets:
+            if(p_net.pin != None):
+                output.append("add_port_state {g} \\".format(g=p_net.name))
+                output.append("\t-state {{default {v}}}".format(v=vdd.value))
+        for g_net in ground_nets:
+            if(g_net.pin != None):
+                output.append("add_port_state {g} \\".format(g=g_net.name))
+                output.append("\t-state {default off}")
```
Comment (harrisonliew) [inline hammer/vlsi/hammer_vlsi_impl.py:2098]: I don't think this is correct according to my understanding of the command... instead it should be a voltage of 0. "off" is distinctly a different state?

File: hammer/vlsi/hammer_vlsi_impl.py [hunk 2]
```diff
@@ -2059,11 +2059,52 @@ def sdf_file(self, value: Optional[str]) -> None:
...
+        #Create Power State Table
+        output.append("create_pst pwr_state_table \\")
+        output.append("\t-supplies {{{p} {g}}}".format(p=" ".join(map(lambda x: x.name, power_nets)),g=" ".join(map(lambda x: x.name, ground_nets)) ))
+        #Add Power States
+        output.append("add_pst_state {d} \\".format(d="aon"))
+        output.append("\t-pst {pwr_state_table} \\")
+        output.append("\t-state {{{s}}}".format(s=" ".join(map(lambda x: "default", power_nets+ground_nets))))
```
Comment (harrisonliew) [inline hammer/vlsi/hammer_vlsi_impl.py:2105]: This map doesn't take into account the names of the nets (get `{default default}`). Might want to do that (including above when the state name is made) for clarity.

**Review by harrisonliew — CHANGES_REQUESTED** [review]
Thanks for this PR! A couple comments:

- It seems like these are UPF 2.0 commands. You might want to include a `upf_version` command in the header. Similarly, [this](https://github.com/ucb-bar/hammer/blob/0ec427c5a8f609b061c8c111d7797014e56c7d07/hammer/config/defaults.yml#L297) comment may need to reflect that
- Building on previous point, supporting UPF 3+ could be a switch? That option there could be like `upf_2.0` or `upf_3.1`, etc. if you desire.
- Is there any corresponding updates needed for the Cadence plugins?

**Comment by kenhoberkeley** [comment]
Really nitpicky here but could we switch to f-strings.  Might just be nice to stick for future code.

File: hammer/vlsi/hammer_vlsi_impl.py (hunk 1 above)
Comment (derrickqi2003) [reply hammer/vlsi/hammer_vlsi_impl.py:2098]: Screen Shot 2023-03-30 at 5 49 42 PM
It appears to be a special exception according to the 2018 manual, but I will change it to be safe.

File: hammer/vlsi/hammer_vlsi_impl.py (hunk 2 above)
Comment (derrickqi2003) [reply hammer/vlsi/hammer_vlsi_impl.py:2105]: Screen Shot 2023-03-30 at 6 00 57 PM
This is the intended power state table generated by those upf commands.
I believe the names of the nets are accounted for in the create_pst_state table. I intended for this to mimic the nominal condition in CPF, thus why there's no specificity. Let me know if I should do it though.

**Comment by derrickqi2003** [comment]
@kenhoberkeley Added fstrings
@harrisonliew Made the voltage change from "off" to 0.0 & updated the documentation to say IEEE 1801-2009

**Pull request merged**